driver-core/src/main/com/mongodb/internal/connection/SingleServerCluster.java

```diff
@@ -41,7 +44,7 @@
 public final class SingleServerCluster extends BaseCluster {
     private static final Logger LOGGER = Loggers.getLogger("cluster");
 
-    private final ClusterableServer server;
+    private final AtomicReference<ClusterableServer> server;
```

Comment:
Introducing this indirection is the only way to do the same thing that was done via the `synchronized` block below. This code weirdness stems from the approach when constructors do mo work than purely constructing objects.

Reply:
I don't understand what the difference is between withLock and synchronized that requires this change.  Can you explain a bit more?

Reply:
It is impossible to assign a value to `SingleServerCluster.server` from within a lambda expression, but was possible from a `synchronized` block inside the body of a constructor.

Reply:
It also may seem like I could have just made the `server` field non-`final` instead of making it `final AtomicReference<ClusterableServer>`. However it is not the case:

- `AtomicReference` is needed only as a mutable container for `ClusterableServer`. The fact that it stores its field internally as `volatile` has no significance to us.
- `final` makes sure that even if a reference to `SingleServerCluster` is published via a data race, `SingleServerCluster.connect` (and other methods reading the `server` field) are guaranteed to observe this it properly initialized: not only `server` is not `null`, but also `server.get()` is guaranteed to return non-`null`.

Reply:
Got it.  Thanks.